### DISCUSSION TITLE
Auth Collection Custom Model always updates email field with nil

### DISCUSSION THREAD

**Discussion by pnmcosta** [pocketbase/pocketbase#3821] [Q&A] [answered]
Hi,

Am having a bit of an issue using Custom Model for a Auth Collection that I do want a read-only/optional email field (oauth only):

![image](https://github.com/pocketbase/pocketbase/assets/550684/014c5003-0e33-45f0-b2a9-0075e92847e2)

Field is populated fine from provider (facebook in this case) but a subsequent update (not on same request) using a custom model save is nilling the email field.

Not sure if this is expected behaviour, I know we have to use `forms`/sensitive fields to be able to create/update these, is there anyway to ensure a custom model field is read only? This is my simplified struct:

```
type Creator struct {
	pb.BaseModel

	Email        string `db:"email" json:"email"`
}
```

Thank you,

**Comment by ganigeorgiev** [comment]
> but a subsequent update (not on same request) using a custom model save is nilling the email field.

I don't understand what do you mean by this.

> ... is there anyway to ensure a custom model field is read only

There is no way in Go to mark a struct field as readonly. Depending on your use case one possible solution could be to define a new `string` type and handle the special behavior by attaching methods to it and/or implementing interfaces like `driver.Valuer`, `json.Marshaler`, etc. but this will really depend on what actually you are doing.

Could you elaborate and/or provide a code sample?

**Reply by pnmcosta** [reply]
I use custom models on all my routes, and after oauth (record create) I ask the user some additional details and save (record update) using a custom route.

Something like this:

```
package main

import (
	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
	pb "github.com/pocketbase/pocketbase/models"
)

var _ pb.Model = (*CustomModel)(nil)

type CustomModel struct {
	pb.BaseModel

	Email string `db:"email" json:"email"`

	AnotherField string `db:"anotherField" json:"anotherField"`
}

func (m *CustomModel) TableName() string {
	return "authcollname" // the name of your collection
}

func myRoute(app core.App, id string) error {
	dao := app.Dao()
	model := &CustomModel{}

	if err := dao.ModelQuery(model).
		AndWhere(dbx.HashExp{"id": id}).
		Limit(1).
		One(model); err != nil {
		return err
	}

	model.AnotherField = "updateme"

	// this is nilling email column
	return dao.Save(model)
}
```

If you run `myRoute` after creating the record you'll see what I mean.

**Reply by pnmcosta** [reply]
> There is no way in Go to mark a struct field as readonly

That's not what I'm looking for, I just want `email` to be ignored when updating custom model with `dao.Save`.

**Reply by pnmcosta** [reply]
The only way I can do it for now is not to have Email as a field of my custom model struct, but it would be nice to read it too. 

Oh wait, can I implement an interface on the custom model type itself? I could definitely live with something like:

```
type CustomModel struct {
	pb.BaseModel

	Email string `json:"email"`

	AnotherField string `db:"anotherField" json:"anotherField"`
}

func (m *CustomModel) Scan(field, value)  {
	if fiels == "email"{
		m.Email = value
	}
}
```

(very much pseudo code, I haven't looked at the interfaces yet, just a thought)

notice the missing db tag on email field.

**Reply by ganigeorgiev** [reply, marked as answer]
I'm not able to reproduce your example. Here is mine:

```go
package main

import (
	"log"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/models"
)

type User struct {
	models.BaseModel

	Email string `db:"email" json:"email"`
	Name  string `db:"name" json:"name"`
}

func (m User) TableName() string {
	return "users"
}

func main() {
	app := pocketbase.New()

	app.OnAfterBootstrap().Add(func(e *core.BootstrapEvent) error {
		model := &User{}

		err := app.Dao().ModelQuery(model).
			AndWhere(dbx.HashExp{"email": "test@example.com"}).
			Limit(1).
			One(model)
		if err != nil {
			return err
		}

		model.Name = "update!"

		return app.Dao().Save(model)
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

The only way I can think the email field will be resetted on `Save()` would be if it wasn't populated in the model in the first place and in this case the struct value will be empty string (eg. if you have custom select clause).

**Reply by pnmcosta** [reply]
Ah yes, my sincere apologies, you are correct, I should have triple checked! I misidentified the root cause 😞 

It's actually my mapping method from a `FindAuthRecordByToken` record to the custom model that was not `record.IgnoreEmailVisibility(true)` and I missed it!

Many thanks for the help 🙇
